src/grammar_optimiser/parse_input.jl

```diff
@@ -0,0 +1,127 @@
+using DataStructures;
+
+"""
+    parse_number(start_index::Int64, input::String)
+
+A helper method that parses a number from a string.
+# Arguments
+- `start_index::Int64`: the index to start parsing from
+- `input::String`: the input string
+# Result
+- `number::String`: the parsed number
+- `i::Int64`: the index of the last character parsed
+"""
+function parse_number(start_index, input)
+    number = ""
+    i = start_index
+    while i <= length(input)
+        char = input[i]
+        if isdigit(char)
+            number = number * char
+        else
+            break
+        end
+        i += 1
+    end
+    return number, i - 1
+end
+
+"""
+    parse_tree(input::String, global_dict::Dict, start_index::Int64)
+
+Parses a tree from a string.
```

Comment:
It is not clear for me what this function does. What kind of tree is parsed here? Shouldn't the output type then be `RuleNode`? 

Could you also add the return types?